e2e/test/DeviceClientX509AuthenticationE2ETests.cs

```diff
@@ -0,0 +1,160 @@
+        }
+
+        [TestMethod]
+        public async Task X509_InvalidDeviceId_Twice_Amqp()
+        {
+            await X509InvalidDeviceIdOpenAsyncTwiceTest(Client.TransportType.Amqp).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task X509_InvalidDeviceId_Twice_Amqp_TCP()
+        {
+            await X509InvalidDeviceIdOpenAsyncTwiceTest(Client.TransportType.Amqp_Tcp_Only).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task X509_InvalidDeviceId_Twice_Amqp_WebSocket()
+        {
+            await X509InvalidDeviceIdOpenAsyncTwiceTest(Client.TransportType.Amqp_WebSocket_Only).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task X509_InvalidDeviceId_Twice_Mqtt()
+        {
+            await X509InvalidDeviceIdOpenAsyncTwiceTest(Client.TransportType.Mqtt).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task X509_InvalidDeviceId_Twice_Mqtt_Tcp()
+        {
+            await X509InvalidDeviceIdOpenAsyncTwiceTest(Client.TransportType.Mqtt_Tcp_Only).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task X509_InvalidDeviceId_Twice_Mqtt_WebSocket()
+        {
+            await X509InvalidDeviceIdOpenAsyncTwiceTest(Client.TransportType.Mqtt_WebSocket_Only).ConfigureAwait(false);
+        }
+
+        private async Task X509InvalidDeviceIdOpenAsyncTest(Client.TransportType transportType)
+        {
+            var deviceClient = CreateDeviceClientWithInvalidId(transportType);
+            using (deviceClient)
+            {
+                try
+                {
+                    await deviceClient.OpenAsync().ConfigureAwait(false);
+                    Assert.Fail("Should throw UnauthorizedException but didn't.");
+                }
+                catch (UnauthorizedException)
+                {
+                    // It should always throw UnauthorizedException
+                }
+
+                // Check TCP connection to verify there is no connection leak
+                // netstat -na | find "[Your Hub IP]" | find "ESTABLISHED" 
+                await Task.Delay(TimeSpan.FromSeconds(10)).ConfigureAwait(false);
+            }
+        }
+
+        private async Task X509InvalidDeviceIdOpenAsyncTwiceTest(Client.TransportType transportType)
+        {
+            var deviceClient = CreateDeviceClientWithInvalidId(transportType);
+            using (deviceClient)
+            {
+                for (int i = 0; i < 2; i++)
+                {
+                    try
+                    {
+                        await deviceClient.OpenAsync().ConfigureAwait(false);
+                        Assert.Fail("Should throw UnauthorizedException but didn't.");
+                    }
+                    catch (UnauthorizedException)
+                    {
+                        // It should always throw UnauthorizedException
+                    }
+                }
```

Comment:
redundant code in task above, can be extracted. Also can you explain the reason of doing it twice? probably hard code value 2 if it makes more sense readability wise.